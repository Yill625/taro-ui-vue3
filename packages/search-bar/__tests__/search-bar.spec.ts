import {
  genMountFn,
  testPropClassAndStyle
} from '@taro-ui-vue3/test-utils/helper'
import { capitalize } from '@vue/shared'
import { ref } from 'vue'
import AtSearchBar from '../index.vue'

const mountFn = genMountFn(AtSearchBar)

describe('AtSearchBar', () => {
  testPropClassAndStyle(mountFn)

  it('should render default and match snapshot', () => {
    const wrapper = mountFn()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop modelValue', () => {
    const wrapper = mountFn({ modelValue: 'value' })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop placeholder', () => {
    const wrapper = mountFn({ placeholder: 'placeholder' })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop maxLength', () => {
    const wrapper = mountFn({ maxLength: 120 })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop fixed', () => {
    const wrapper = mountFn({ fixed: true })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop focus', () => {
    const wrapper = mountFn({ fucos: true })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop disabled', () => {
    const wrapper = mountFn({ disabled: true })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop showActionButton', () => {
    const wrapper = mountFn({ showActionButton: true })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop actionName', () => {
    const wrapper = mountFn({ actionName: 'actionName' })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop inputType number', () => {
    const wrapper = mountFn({ inputType: 'number' })
    expect(wrapper.element).toMatchSnapshot()
  })
})

describe('AtSearchBar events', () => {
  it('should emit update:modelValue', async () => {
    const modelValue = ref('props.value')
    const onUpdateValue = jest.fn((e) => modelValue.value = e)
    const wrapper = mountFn({
      modelValue,
      ['onUpdate:modelValue']: onUpdateValue
    })

    await wrapper
      .find('.at-search-bar .at-search-bar__input')
      .trigger('input', { detail: { value: 'value' } })

    expect(wrapper.emitted()).toHaveProperty('update:modelValue')
    expect(wrapper.emitted('update:modelValue')![0]).toEqual(['value'])
    expect(onUpdateValue).toBeCalled()
    expect(modelValue.value).toEqual('value')

    await wrapper
      .find('.at-search-bar .at-search-bar__clear')
      .trigger('touchstart')

    await wrapper.vm.$nextTick()

    expect(modelValue.value).toEqual('')
  })

  it.each([
    'focus', 'blur', 'confirm'
  ])('should emit %s', async (eventName) => {
    const mockFn = jest.fn()
    const wrapper = mountFn({ [`on${capitalize(eventName)}`]: mockFn })

    await wrapper.find('.at-search-bar .at-search-bar__input').trigger(eventName)

    expect(wrapper.emitted()).toHaveProperty(eventName)
    expect(mockFn).toBeCalled()
  })

  it('should emit ActionClick', async () => {
    const onActionClick = jest.fn()
    const wrapper = mountFn({ onActionClick })

    await wrapper.find('.at-search-bar .at-search-bar__action').trigger('tap')

    expect(wrapper.emitted()).toHaveProperty('action-click')
    expect(onActionClick).toBeCalled()
  })

  // @TODO: remove onClear ?!
  it('should trigger onClear', async () => {
    const onClear = jest.fn()
    const wrapper = mountFn({ onClear })

    const clearElement = wrapper.find('.at-search-bar .at-search-bar__clear')
    expect(clearElement.exists()).toBe(false)

    await wrapper.setProps({ modelValue: 'props.modelValue is present now' })
    await wrapper.vm.$nextTick()

    const clearElement2 = wrapper.find('.at-search-bar .at-search-bar__clear')
    expect(clearElement2.exists()).toBe(true)

    await clearElement2.trigger('touchstart')

    expect(onClear).toBeCalled()
  })
})
