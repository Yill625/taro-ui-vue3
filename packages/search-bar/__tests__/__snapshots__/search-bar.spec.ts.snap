// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtSearchBar should render default AtSearchBar 1`] = `
<view
  class="at-search-bar"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 1;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: visible;"
      >
        搜索
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      id="weui-input_2020"
      maxlength="140"
      type="text"
    />
    
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 0; margin-right: -59px;"
  >
    搜索
  </view>
</view>
`;

exports[`AtSearchBar should render prop -- actionName 1`] = `
<view
  class="at-search-bar"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 1;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: visible;"
      >
        搜索
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      id="weui-input_2020"
      maxlength="140"
      type="text"
    />
    
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 0; margin-right: -171px;"
  >
    actionName
  </view>
</view>
`;

exports[`AtSearchBar should render prop -- class 1`] = `
<view
  class="test at-search-bar test"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 1;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: visible;"
      >
        搜索
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      id="weui-input_2020"
      maxlength="140"
      type="text"
    />
    
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 0; margin-right: -59px;"
  >
    搜索
  </view>
</view>
`;

exports[`AtSearchBar should render prop -- disabled 1`] = `
<view
  class="at-search-bar"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 1;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: visible;"
      >
        搜索
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      disabled=""
      id="weui-input_2020"
      maxlength="140"
      type="text"
    />
    
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 0; margin-right: -59px;"
  >
    搜索
  </view>
</view>
`;

exports[`AtSearchBar should render prop -- fixed 1`] = `
<view
  class="at-search-bar at-search-bar--fixed"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 1;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: visible;"
      >
        搜索
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      id="weui-input_2020"
      maxlength="140"
      type="text"
    />
    
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 0; margin-right: -59px;"
  >
    搜索
  </view>
</view>
`;

exports[`AtSearchBar should render prop -- focus 1`] = `
<view
  class="at-search-bar"
  fucos="true"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 1;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: visible;"
      >
        搜索
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      id="weui-input_2020"
      maxlength="140"
      type="text"
    />
    
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 0; margin-right: -59px;"
  >
    搜索
  </view>
</view>
`;

exports[`AtSearchBar should render prop -- inputType number 1`] = `
<view
  class="at-search-bar"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 1;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: visible;"
      >
        搜索
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      id="weui-input_2020"
      maxlength="140"
      type="number"
    />
    
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 0; margin-right: -59px;"
  >
    搜索
  </view>
</view>
`;

exports[`AtSearchBar should render prop -- maxLength 1`] = `
<view
  class="at-search-bar"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 1;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: visible;"
      >
        搜索
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      id="weui-input_2020"
      maxlength="120"
      type="text"
    />
    
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 0; margin-right: -59px;"
  >
    搜索
  </view>
</view>
`;

exports[`AtSearchBar should render prop -- placeholder 1`] = `
<view
  class="at-search-bar"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 1;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: visible;"
      >
        placeholder
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      id="weui-input_2020"
      maxlength="140"
      type="text"
    />
    
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 0; margin-right: -59px;"
  >
    搜索
  </view>
</view>
`;

exports[`AtSearchBar should render prop -- showActionButton 1`] = `
<view
  class="at-search-bar"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 1;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: visible;"
      >
        搜索
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      id="weui-input_2020"
      maxlength="140"
      type="text"
    />
    
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 1; margin-right: 0px;"
  >
    搜索
  </view>
</view>
`;

exports[`AtSearchBar should render prop -- style 1`] = `
<view
  class="at-search-bar"
  style="color: red;"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 1;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: visible;"
      >
        搜索
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      id="weui-input_2020"
      maxlength="140"
      type="text"
    />
    
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 0; margin-right: -59px;"
  >
    搜索
  </view>
</view>
`;

exports[`AtSearchBar should render prop -- value 1`] = `
<view
  class="at-search-bar"
>
  <view
    class="at-search-bar__input-cnt"
  >
    <view
      class="at-search-bar__placeholder-wrap"
      style="flex-grow: 0;"
    >
      <text
        class="at-icon at-icon-search"
      />
      <text
        class="at-search-bar__placeholder"
        style="visibility: hidden;"
      >
        搜索
      </text>
    </view>
    <input
      class="at-search-bar__input"
      confirmtype="search"
      id="weui-input_2020"
      maxlength="140"
      type="text"
    />
    <view
      class="at-search-bar__clear"
      style="display: flex;"
    >
      <text
        class="at-icon at-icon-close-circle"
      />
    </view>
  </view>
  <view
    class="at-search-bar__action"
    style="opacity: 1; margin-right: 0px;"
  >
    搜索
  </view>
</view>
`;
